react/states_and_effects/more_on_state.md

```diff
@@ -0,0 +1,213 @@
+### Introduction
+
+This lesson will build upon the previous lesson on state and introduce further concepts and techniques to manage state. Let's dive in.
+
+### Lesson overview
+
+This section contains a general overview of topics that you will learn in this lesson.
+
+- How to structure state.
+- How state updates.
+- Learn about controlled components.
+
+### How to structure state
+
+Managing and structuring state effectively is by far one of the most crucial parts of building your application. If not done correctly, it can become a source of bugs and headaches.   
+
+The assignment items go through the topic thoroughly, but as a general rule of thumb: don't put values in state that can be calculated using existing values, state, and/or props. 
+
+#### State is immutable
+
+<span id="state-is-immutable"></span>Mutating state is no-go area in React as it leads to unpredictable results. Primitives are already immutable, but if you are using reference type values i.e. arrays and objects, never mutate them. Instead always use the `setState` function to update state. 
+
+Make sure to run the following example locally and see the difference for yourself.
+
+~~~jsx
+function Person() {
+   const [person, setPerson] = useState({ name: 'John', age: 100 });
+
+   // BAD - Don't do this!
+   const handleIncreaseAge = () =>{
+      // mutating the current state object
+      person.age = person.age + 1;
+      setPerson(person);
+   }
+
+   // GOOD - Do this!
+   const handleIncreaseAge = () =>{
+      // copy the existing person object into a new object 
+      // while updating the age property
+      const newPerson = { ...person, age: person.age + 1 };
+      setPerson(newPerson);
+   }
+   
+   return (
+      <>
+         <h1>{person.name}</h1>
+         <h2>{person.age}</h2>
+         <button onClick={handleIncreaseAge}>Increase age</button>
+      </>
+   )
+}
+~~~
+
+<div class="lesson-note lesson-note--tip" markdown="1">
+
+#### Nested objects and arrays in state
+
+In the above example, notice how we had to copy the existing state object. Herein lies another lesson, nested objects and arrays as state can get tricky fast, since you will have to replicate the nested items as well. Be careful when using them.
```

Comment:
```suggestion
In the above example, notice how we had to copy the existing state object. Herein lies another lesson, nested objects and arrays as state can get tricky fast since you will have to replicate the nested items as well. Be careful when using them.
```
removed comma